gandalf.js

```diff
@@ -1,54 +1,55 @@
 /**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?